lib/SILOptimizer/Transforms/ARCCodeMotion.cpp

```diff
@@ -375,16 +377,50 @@ bool RetainCodeMotionContext::requireIteration() {
 
 void RetainCodeMotionContext::initializeCodeMotionDataFlow() {
   // Find all the RC roots in the function.
+  llvm::DenseMap<SILValue, Atomicity> AtomicityOfRoot;
+
   for (auto &BB : *F) {
     for (auto &II : BB) {
+      auto RCInst = dyn_cast<RefCountingInst>(&II);
+      if (!RCInst)
+        continue;
+      // Remember the atomicity of the root. If any
+      // of the RC operations on a the root are atomic,
+      // then the root is considered to be atomic.
+      Atomicity InstAtomicity = (RCInst->getAtomicity() == Atomicity::Atomic)
+                                    ? Atomicity::Atomic
+                                    : Atomicity::NonAtomic;
+      //auto Root = getRCRoot(&II);
+      auto Root = getRCRoot(RCInst->getOperand(0));
+
+      auto newRoot = AtomicityOfRoot.count(Root);
+      auto &RootAtomicity = AtomicityOfRoot[Root];
+
+      if (!newRoot)
```

Comment:
Instead of looking up Root in the (not needed) AtomicityOfRoot, you can use this condition to check if you first see this root:
  if (RCRootIndex.find(Root) != RCRootIndex.end()) {